src/TwigYard/Middleware/Data/DataMiddleware.php

```diff
@@ -12,46 +16,103 @@
 
 class DataMiddleware implements MiddlewareInterface
 {
+    const TYPE_LOCAL = 'local';
+    const TYPE_HTTP = 'http';
+
+    const FORMAT_YAML = 'yml';
+    const FORMAT_JSON = 'json';
+
     /**
      * @var string
      */
     private $dataDir;
+
     /**
      * @var \TwigYard\Component\AppState
      */
     private $appState;
 
     /**
+     * @var \TwigYard\Component\CurlDownloader
+     */
+    private $curlDownloader;
+
+    /**
      * @param \TwigYard\Component\AppState $appState
      * @param string $dataDir
+     * @param \TwigYard\Component\CurlDownloader $curlDownloader
      */
-    public function __construct(AppState $appState, $dataDir)
+    public function __construct(AppState $appState, $dataDir, CurlDownloader $curlDownloader)
     {
         $this->dataDir = $dataDir;
         $this->appState = $appState;
+        $this->curlDownloader = $curlDownloader;
     }
 
     /**
      * @param \Psr\Http\Message\ServerRequestInterface $request
      * @param \Psr\Http\Message\ResponseInterface $response
-     * @param callable|\TwigYard\Middleware\MiddlewareInterface $next
-     * @return \Psr\Http\Message\ResponseInterface $response
+     * @param callable $next
+     * @throws \TwigYard\Exception\CannotAccessRemoteSourceException
+     * @throws \TwigYard\Exception\InvalidDataFormatException
+     * @throws \TwigYard\Exception\InvalidDataTypeException
      * @throws \TwigYard\Exception\InvalidSiteConfigException
+     * @return \Psr\Http\Message\ResponseInterface $response
      */
     public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next)
     {
         if (array_key_exists('data', $this->appState->getConfig())) {
             $conf = $this->appState->getConfig()['data'];
             $data = [];
-            foreach ($conf as $var => $file) {
-                $dataFile = $this->appState->getSiteDir() . '/' . $this->dataDir . '/' . $file;
-                if (!is_file($dataFile)) {
-                    throw new InvalidSiteConfigException(sprintf('Missing data file: %s', $dataFile));
+            foreach ($conf as $var => $source) {
+                if (!is_array($source)) {
+                    $source = [
+                        'type' => self::TYPE_LOCAL,
+                        'format' => self::FORMAT_YAML,
+                        'path' => $source,
+                    ];
+                }
+                if (isset($source['url'])) {
```

Comment:
I this necessaery? Setting `url` as `path` seems sort of hacky.

Reply:
I suppose more data channels in the future not only http and local file and that's why I wanted to unify keyname of source location, however in the config it is more intuitive to write "url" for http type of resource and "path" for local file resource. Maybe it is not a properly result and I have a second option on my mind. Create a more clever method / component, which converts source configuration to unified structure, validate it and then this one will be processed.
For instance:
Http source in json format
data:
    resource:
        type: http
        format: json
        url: http://www.example.com/resource.json
will be validated and in case of missing some mandatory key or malformed values for specific resource type throws an error immediatelly. If the configuration will be ok, the method returns specific configured Downloader (CurlDownloader with SSL verification enabled in case of https, LocalFileReader for local files, etc. with defined Interface) and this one will be used for retrieving of the content. This solution is a little bit more complex and I wanted to implement the feature quick and simple. What do you mean? Should I implement rather the second more complex solution with the view to the future?

Reply:
1. Could we call it `resource` and we can use it for both remote url and local file path?
2. I would prefer not to not add any more complexity now. Maybe we could create a private `getContent()` method and break up the long `__invoke()` method a bit. I think that should be enough.